README.md
# Very short description of the package

[![Latest Version on Packagist](https://img.shields.io/packagist/v/kingsleyudenewu/estoresms.svg?style=flat-square)](https://packagist.org/packages/kingsleyudenewu/estoresms)
[![Build Status](https://img.shields.io/travis/kingsleyudenewu/estoresms/master.svg?style=flat-square)](https://travis-ci.org/kingsleyudenewu/estoresms)
[![Quality Score](https://img.shields.io/scrutinizer/g/kingsleyudenewu/estoresms.svg?style=flat-square)](https://scrutinizer-ci.com/g/kingsleyudenewu/estoresms)
[![Total Downloads](https://img.shields.io/packagist/dt/kingsleyudenewu/estoresms.svg?style=flat-square)](https://packagist.org/packages/kingsleyudenewu/estoresms)

This is where your description should go. Try and limit it to a paragraph or two, and maybe throw in a mention of what PSRs you support to avoid any confusion with users and contributors.

## Installation

You can install the package via composer:

```bash
composer require kingsleyudenewu/estoresms
```

## Usage

``` php
// Usage description here
```

### Testing

``` bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email kingzpacking@gmail.com instead of using the issue tracker.

## Credits

- [Udenewu Kingsley](https://github.com/kingsleyudenewu)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## Laravel Package Boilerplate

This package was generated using the [Laravel Package Boilerplate](https://laravelpackageboilerplate.com).

CHANGELOG.md
# Changelog

All notable changes to `estoresms` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/ExampleTest.php
<?php

namespace Kingsleyudenewu\Estoresms\Tests;

use Orchestra\Testbench\TestCase;
use Kingsleyudenewu\Estoresms\EstoresmsServiceProvider;

class ExampleTest extends TestCase
{

    protected function getPackageProviders($app)
    {
        return [EstoresmsServiceProvider::class];
    }
    
    /** @test */
    public function true_is_true()
    {
        $this->assertTrue(true);
    }
}


src/Estoresms.php
<?php

namespace Kingsleyudenewu\Estoresms;

use Kingsleyudenewu\Estoresms\traits\SmsResponse;
use Kingsleyudenewu\Estoresms\traits\Utils;

class Estoresms
{
    use Utils;

    public function accountBalance()
    {
        $balance = $this->getRequest(config('estoresms.url') . 'smsapi.php?' .
            'username=' . config('estoresms.username') .
            '&password=' . config('estoresms.password') .
            '&balance=true&');

        if (!is_null($balance) || !empty($balance)) {
            // Check if its response is not an array od error codes
            if (in_array($balance, $this->getErrorResponses(config('estoresms.sms_response'),
                config('estoresms.sms_response.OK')))) {
                return $this->setResponse(false,
                    null,
                    $this->getConfigResponseMessage(config('estoresms.sms_response'), $balance)
                );
            }
            return $this->setResponse(true, $balance, 'success');
        }
        return $this->setResponse(false, null, 'error');
    }

    public function sendSms($data = [])
    {
        if (empty($data) || !is_array($data))
        {
            return $this->setResponse(false, null, 'Invalid payload');
        }
        if (is_null($data['sender']) || empty($data['sender']))
        {
            return $this->setResponse(false, null, $this->getConfigResponseMessage(config('estoresms.sms_response'), '-2914'));
        }

        if (is_null($data['recipient']) || empty($data['recipient']) || !is_string($data['recipient']))
        {
            return $this->setResponse(false, null, $this->getConfigResponseMessage(config('estoresms.sms_response'), '-2912'));
        }

        if (is_null($data['message']) || empty($data['message']) || !is_string($data['message']))
        {
            return $this->setResponse(false, null, $this->getConfigResponseMessage(config('estoresms.sms_response'), '-2913'));
        }

        if ($data['sms_type'] == 'flash')
        {
            if ($data['dnd'])
            {
                $send_sms = $this->getRequest(config('estoresms.url').
                    'flashsms_api.php?username='. config('estoresms.username').
                    '&password='.config('estoresms.username').
                    '&recipient='.$data['recipient'].
                    '&message='.$data['message'].
                    '&dnd='.$data['dnd']
                );
            }
            else
            {
                $send_sms = $this->getRequest(config('estoresms.url').
                    'flashsms_api.php?username='. config('estoresms.username').
                    '&password='.config('estoresms.username').
                    '&recipient='.$data['recipient'].
                    '&message='.$data['message'].'&'
                );
            }
        }
        elseif ($data['sms_type'] == 'sms')
        {
            if ($data['dnd'])
            {
                $send_sms = $this->getRequest(config('estoresms.url').
                    'smsapi.php?username='. config('estoresms.username').
                    '&password='.config('estoresms.username').
                    '&recipient='.$data['recipient'].
                    '&message='.$data['message'].
                    '&dnd='.$data['dnd']
                );
            }
            else
            {
                $send_sms = $this->getRequest(config('estoresms.url').
                    'smsapi.php?username='. config('estoresms.username').
                    '&password='.config('estoresms.username').
                    '&recipient='.$data['recipient'].
                    '&message='.$data['message'].'&'
                );
            }
        }

        // Check if we got an error response from the SMS API
        $error_response = $this->displayErrorMessage(config('estoresms.sms_response'), $send_sms);
        if (!is_null($error_response)){
            return $error_response;
        }

        // Return success message and the number of SMS unit used
        return $this->setResponse(true, $send_sms, 'success');
    }

    public function getBillPaymentList($category)
    {
        $hash = hash('sha512', config('estoresms.token').config('estoresms.email').config('estoresms.username') );

        // Fetch product list
        $payload = [
          'username' => config('estoresms.username'),
          'hash' => $hash,
          'category' => $category,
        ];
        $product_list = $this->postRequest(config('url').'bill_payment_processing/v/1/', $payload, []);
        if (optional($product_list)->response == 'OK')
        {
            return $product_list;
        }
    }

    public function getProductPlans()
    {

    }
}


src/traits/Utils.php
<?php


namespace Kingsleyudenewu\Estoresms\traits;


use Illuminate\Support\Arr;
use Ixudra\Curl\Facades\Curl;

trait Utils
{
    private function getRequest($url, $header = [])
    {
        $response = Curl::to($url)
            ->withHeaders($header)
            ->asJson()
            ->get();

        if (!is_null($response))return $response;
        return false;
    }

    private function postRequest($url, $data, $header=[])
    {
        $response = Curl::to($url)
            ->withHeaders($header)
            ->withData($data)
            ->asJson()
            ->post();

        if (!is_null($response))return $response;
        return false;
    }

    private function setResponse($status, $data, $message)
    {
        $response = [];
        $response['status'] = $status;
        if (!is_null($data))
        {
            $response['data'] = $data;
        }

        $response['message'] = $message;
        return response()->json($response);
    }

    private function getErrorResponses($response_type, $response_filter_value)
    {
        $filtered = Arr::except($response_type, $response_filter_value);
        [$keys, $values] = Arr::divide($filtered);
        return $keys;
    }

    private function getConfigResponseMessage($response_type, $key)
    {
        return $response_type[$key];
    }

    private function displayErrorMessage($response_type, $error_code)
    {
        if (in_array($error_code, $this->getErrorResponses($response_type, $response_type['OK'])))
        {
            return $this->setResponse(
                false,
                null,
                $this->getConfigResponseMessage($response_type, $error_code)
            );
        }

        return null;
    }

    private function bill_payment_api_validation($category, $product)
    {
        $hash = hash('sha512', config('estoresms.token').config('estoresms.email').config('estoresms.username') );
        $payload = [
            'username' => config('estoresms.username'),
            'hash' => $hash,
            'category' => $category,
            'product' => $product,
        ];
        $product_list = $this->postRequest(config('url').'bill_payment_processing/v/1/', $payload, []);
    }
}


src/EstoresmsServiceProvider.php
<?php

namespace Kingsleyudenewu\Estoresms;

use Illuminate\Support\ServiceProvider;

class EstoresmsServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        /*
         * Optional methods to load your package assets
         */
        // $this->loadTranslationsFrom(__DIR__.'/../resources/lang', 'estoresms');
        // $this->loadViewsFrom(__DIR__.'/../resources/views', 'estoresms');
        // $this->loadMigrationsFrom(__DIR__.'/../database/migrations');
        // $this->loadRoutesFrom(__DIR__.'/routes.php');

        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__.'/../config/config.php' => config_path('estoresms.php'),
            ], 'config');

            // Publishing the views.
            /*$this->publishes([
                __DIR__.'/../resources/views' => resource_path('views/vendor/estoresms'),
            ], 'views');*/

            // Publishing assets.
            /*$this->publishes([
                __DIR__.'/../resources/assets' => public_path('vendor/estoresms'),
            ], 'assets');*/

            // Publishing the translation files.
            /*$this->publishes([
                __DIR__.'/../resources/lang' => resource_path('lang/vendor/estoresms'),
            ], 'lang');*/

            // Registering package commands.
            // $this->commands([]);
        }
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        // Automatically apply the package configuration
        $this->mergeConfigFrom(__DIR__.'/../config/config.php', 'estoresms');

        // Register the main class to use with the facade
        $this->app->singleton('estoresms', function () {
            return new Estoresms;
        });
    }
}


config/config.php
<?php

/*
 * You can place your custom package configuration in here.
 */
return [
    'sms_response' => [
        'OK' => 'Successful',
        '-2904' => 'SMS sending failed',
        '-2905' => 'Invalid username/password combination',
        '-2906' => 'Credit exhausted',
        '-2907' => 'Gateway unavailable',
        '-2908' => 'Invalid schedule date format',
        '-2909' => 'Unable to schedule',
        '-2910' => 'Username is empty',
        '-2911' => 'Password is empty',
        '-2912' => 'Recipient is empty',
        '-2913' => 'Message is empty',
        '-2914' => 'Sender is empty',
        '-2915' => 'One or more required fields are empty',
        '-2916' => 'Blocked message content',
        '-2917' => 'Blocked sender ID',
    ],
    'bill_payment_category' => [
        'tv',
        'internet',
        'electricity',
        'waec',
    ],

    'bill_payment_response' => [
        'OK' => 'Successful',
        'FAIL' => 'Not Successful',
        'B001' => 'Authentication Failed',
        'B002' => 'Operation not allowed on account',
        'B003' => 'API not enabled for this account',
        'B004' => 'Invalid request',
        'B005' => 'Product has no plan in this category',
        'B006' => 'Validation failed',
        'B007' => 'Insufficient Balance',
        'B008' => 'Invalid Amount',
        'B009' => 'Operation Failed',
    ],

    'airtime_response' => [
        'OK' => 'Successful',
        'FAIL' => 'Not Successful',
        'P000' => 'Incomplete Information',
        'P001' => 'Authentication failed',
        'P002' => 'Invalid Network selected',
        'P003' => 'Phone number is required',
        'P004' => 'Amount is required',
        'P005' => 'Amount is not numeric',
        'P006' => 'Minimum recharge allowed per transaction N1',
        'P007' => 'Invalid phone number',
        'P008' => 'Network selected for phone number is invalid',
        'P009' => 'Insufficient Balance',
        'P010' => 'API Token not enabled',
        'P011' => 'Process Failed',
        'P012' => 'Unable to complete request',
        'P013' => 'Maximum recharge allowed per transaction N50,000.',
        'P014' => 'Maximum number of pins that can be generated at once is 1000.',
    ],

    'token' => env('ESTORE_TOKEN'),
    'email' => env('ESTORE_EMAIL'),
    'username' => env('ESTORE_USERNAME'),
    'password' => env('ESTORE_PASSWORD'),
    'network_list' => 'https://estoresms.com/network_list/v/2',
    'url' => 'https://www.estoresms.com/',

];
